File: src/new.go
package vibe

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
)

const workspacesDir = ".vibe-workspaces"

var filesToCopy = []string{
	".envrc",
	".claude/settings.local.json",
}

func NewTask(taskName string) error {
	sanitized := sanitizeTaskName(taskName)
	if sanitized == "" {
		return fmt.Errorf("invalid task name: %q", taskName)
	}

	gitRoot, err := getGitRoot()
	if err != nil {
		return fmt.Errorf("not a git repository")
	}

	repoName := getRepoName(gitRoot)
	home, err := os.UserHomeDir()
	if err != nil {
		return fmt.Errorf("failed to get home directory: %w", err)
	}

	worktreePath := filepath.Join(home, workspacesDir, repoName, sanitized)
	branchName := fmt.Sprintf("feature/%s", sanitized)

	if err := createWorktree(gitRoot, worktreePath, branchName); err != nil {
		return fmt.Errorf("failed to create worktree: %w", err)
	}

	for _, file := range filesToCopy {
		copyIfExists(filepath.Join(gitRoot, file), filepath.Join(worktreePath, file))
	}

	if !commandExists("direnv") {
		fmt.Fprintln(os.Stderr, "Warning: direnv not installed, skipping direnv allow")
	} else {
		cmd := exec.Command("direnv", "allow")
		cmd.Dir = worktreePath
		cmd.Run()
	}

	if err := addToClaudeConfig(worktreePath); err != nil {
		fmt.Fprintf(os.Stderr, "Warning: failed to update claude config: %v\n", err)
	}

	if !commandExists("claude") {
		fmt.Fprintln(os.Stderr, "Warning: claude not installed")
		return nil
	}

	cmd := exec.Command("claude", fmt.Sprintf("/rename %s", sanitized))
	cmd.Dir = worktreePath
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Run()

	return nil
}

func getGitRoot() (string, error) {
	out, err := exec.Command("git", "rev-parse", "--show-toplevel").Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

func getRepoName(gitRoot string) string {
	out, err := exec.Command("git", "rev-parse", "--git-common-dir").Output()
	if err != nil {
		return ""
	}
	gitDir := strings.TrimSpace(string(out))
	if !filepath.IsAbs(gitDir) {
		gitDir = filepath.Join(gitRoot, gitDir)
	}
	return filepath.Base(filepath.Dir(gitDir))
}

func sanitizeTaskName(name string) string {
	name = strings.TrimSpace(name)
	re := regexp.MustCompile(`[^a-zA-Z0-9_-]+`)
	name = re.ReplaceAllString(name, "-")
	name = strings.Trim(name, "-")
	return name
}

func createWorktree(gitRoot, worktreePath, branchName string) error {
	cmd := exec.Command("git", "worktree", "add", worktreePath, "-b", branchName)
	cmd.Dir = gitRoot
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

func copyIfExists(src, dst string) error {
	data, err := os.ReadFile(src)
	if err != nil {
		return nil
	}
	if err := os.MkdirAll(filepath.Dir(dst), 0755); err != nil {
		return err
	}
	return os.WriteFile(dst, data, 0644)
}

func commandExists(name string) bool {
	_, err := exec.LookPath(name)
	return err == nil
}

func addToClaudeConfig(worktreePath string) error {
	home, err := os.UserHomeDir()
	if err != nil {
		return err
	}

	configPath := filepath.Join(home, ".claude.json")
	data, err := os.ReadFile(configPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}

	var config map[string]interface{}
	if err := json.Unmarshal(data, &config); err != nil {
		return err
	}

	projects, ok := config["projects"].(map[string]interface{})
	if !ok {
		projects = make(map[string]interface{})
		config["projects"] = projects
	}

	if _, exists := projects[worktreePath]; !exists {
		projects[worktreePath] = map[string]interface{}{
			"hasTrustDialogAccepted": true,
		}
	}

	newData, err := json.MarshalIndent(config, "", "  ")
	if err != nil {
		return err
	}

	return os.WriteFile(configPath, newData, 0644)
}

func NewTaskMulti(taskName string) error {
	sanitized := sanitizeTaskName(taskName)
	if sanitized == "" {
		return fmt.Errorf("invalid task name: %q", taskName)
	}

	if !commandExists("ghq") {
		return fmt.Errorf("ghq is not installed")
	}
	if !commandExists("fzf") {
		return fmt.Errorf("fzf is not installed")
	}

	repos, err := selectReposWithFzf()
	if err != nil {
		return err
	}
	if len(repos) == 0 {
		return fmt.Errorf("no repositories selected")
	}

	home, err := os.UserHomeDir()
	if err != nil {
		return fmt.Errorf("failed to get home directory: %w", err)
	}

	var worktreePaths []string
	for _, gitRoot := range repos {
		repoName := getRepoName(gitRoot)
		worktreePath := filepath.Join(home, workspacesDir, repoName, sanitized)
		branchName := fmt.Sprintf("feature/%s", sanitized)

		if err := createWorktree(gitRoot, worktreePath, branchName); err != nil {
			fmt.Fprintf(os.Stderr, "Warning: failed to create worktree for %s: %v\n", repoName, err)
			continue
		}

		for _, file := range filesToCopy {
			copyIfExists(filepath.Join(gitRoot, file), filepath.Join(worktreePath, file))
		}

		if commandExists("direnv") {
			cmd := exec.Command("direnv", "allow")
			cmd.Dir = worktreePath
			cmd.Run()
		}

		if err := addToClaudeConfig(worktreePath); err != nil {
			fmt.Fprintf(os.Stderr, "Warning: failed to update claude config for %s: %v\n", repoName, err)
		}

		worktreePaths = append(worktreePaths, worktreePath)
		fmt.Printf("Created worktree: %s\n", worktreePath)
	}

	if len(worktreePaths) == 0 {
		return fmt.Errorf("no worktrees were created")
	}

	return nil
}

func selectReposWithFzf() ([]string, error) {
	// Get repository list from ghq
	ghqOut, err := exec.Command("ghq", "list", "--full-path").Output()
	if err != nil {
		return nil, fmt.Errorf("failed to run ghq: %w", err)
	}

	// Run fzf with multi-select
	fzfCmd := exec.Command("fzf", "--multi", "--prompt", "Select repositories (TAB to select): ")
	fzfCmd.Stdin = strings.NewReader(string(ghqOut))
	fzfCmd.Stderr = os.Stderr

	// fzf needs access to tty for interactive selection
	tty, err := os.OpenFile("/dev/tty", os.O_RDONLY, 0)
	if err == nil {
		fzfCmd.ExtraFiles = []*os.File{tty}
		defer tty.Close()
	}

	out, err := fzfCmd.Output()
	if err != nil {
		// fzf returns exit code 130 when user presses Ctrl-C, exit code 1 when no selection
		if exitErr, ok := err.(*exec.ExitError); ok {
			if exitErr.ExitCode() == 130 || exitErr.ExitCode() == 1 {
				return nil, nil
			}
		}
		return nil, fmt.Errorf("fzf error: %w", err)
	}

	var selected []string
	for _, line := range strings.Split(strings.TrimSpace(string(out)), "\n") {
		if line = strings.TrimSpace(line); line != "" {
			selected = append(selected, line)
		}
	}

	return selected, nil
}

File: src/new.ts
import { $ } from "bun";
import { homedir } from "os";
import { join } from "path";
import { getGitRoot, createWorktree, getDefaultBranch } from "./git";
import { getGhqRoot } from "./ghq";
import { loadProjectConfig, PROJECT_CONFIG_FILE } from "./project";

export interface RepoConfig {
  path: string;
  branch: string;
  worktreePath: string;
}

export function sanitizeTaskName(task: string): string {
  return task
    .trim()
    .replace(/[^a-zA-Z0-9_-]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

export async function newCommand(
  task: string,
  prefix: string
): Promise<void> {
  const config = await loadProjectConfig(process.cwd());
  if (!config) {
    console.error(`${PROJECT_CONFIG_FILE} not found in current directory.`);
    console.error('Run "vibe create-project" to create one.');
    process.exit(1);
  }

  const taskName = sanitizeTaskName(task);
  if (!taskName) {
    console.error("Invalid task name");
    process.exit(1);
  }

  const gitRoot = await getGitRoot();
  const baseBranch = await getDefaultBranch(gitRoot);
  const ghqRoot = await getGhqRoot();
  const branchName = `${prefix}${taskName}`;

  const additionalRepos: RepoConfig[] = [];

  for (const [ghqPath, repoConfig] of Object.entries(config.repos)) {
    const repoName = ghqPath.split("/").pop() || ghqPath;
    const absolutePath = join(ghqRoot, ghqPath);

    const isGitRepo = (await $`git -C ${absolutePath} rev-parse --git-dir`.nothrow().quiet()).exitCode === 0;
    if (!isGitRepo) {
      console.error(`Repository not found: ${ghqPath}`);
      console.error(`Run "ghq get ${ghqPath}" first.`);
      continue;
    }

    const sourceBranch =
      repoConfig.defaultTarget || (await getDefaultBranch(absolutePath));
    const worktreePath = join(process.cwd(), repoName);

    console.log(`Creating worktree for ${repoName}...`);
    await createWorktree(absolutePath, worktreePath, branchName, sourceBranch);

    await copyWorktreeIncludeFiles(absolutePath, worktreePath);

    if (repoConfig.setupCommand) {
      console.log(`Running setup for ${repoName}...`);
      const result =
        await $`sh -c ${repoConfig.setupCommand}`.cwd(worktreePath).nothrow();
      if (result.exitCode !== 0) {
        console.error(`Warning: setupCommand failed for ${repoName}`);
      }
    }

    await runDirenvAllow(worktreePath);
    await addToClaudeConfig(homedir(), worktreePath);

    additionalRepos.push({
      path: absolutePath,
      branch: sourceBranch,
      worktreePath,
    });
  }

  if (additionalRepos.length === 0) {
    console.error("No repositories were set up.");
    process.exit(1);
  }

  await runClaude(process.cwd(), additionalRepos, baseBranch);

  console.log("\nWorkspaces created:");
  for (const repo of additionalRepos) {
    console.log(`  - ${repo.worktreePath} (branch: ${branchName})`);
  }
}

async function copyWorktreeIncludeFiles(
  repoPath: string,
  worktreePath: string
): Promise<void> {
  const includeFile = Bun.file(join(repoPath, ".worktreeinclude"));
  if (!(await includeFile.exists())) return;

  const content = await includeFile.text();
  const patterns = content
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0 && !line.startsWith("#"));

  if (patterns.length === 0) return;

  const result = await $`git -C ${repoPath} ls-files --others --ignored --exclude-standard -- ${patterns}`
    .nothrow()
    .text();
  const files = result
    .split("\n")
    .filter((f) => f.length > 0);

  for (const file of files) {
    await copyIfExists(join(repoPath, file), join(worktreePath, file));
  }
}

async function copyIfExists(src: string, dst: string): Promise<void> {
  const file = Bun.file(src);
  if (await file.exists()) {
    const dir = dst.substring(0, dst.lastIndexOf("/"));
    if (dir !== dst) {
      await $`mkdir -p ${dir}`.nothrow();
    }
    await Bun.write(dst, file);
  }
}

async function runDirenvAllow(path: string): Promise<void> {
  await $`direnv allow ${path}`.nothrow().quiet();
}

async function runClaude(
  worktreePath: string,
  additionalRepos: RepoConfig[] = [],
  targetBranch?: string
): Promise<void> {
  const args = ["claude"];

  for (const repo of additionalRepos) {
    args.push("--add-dir", repo.worktreePath);
  }

  const env: Record<string, string> = { ...process.env } as Record<
    string,
    string
  >;
  if (additionalRepos.length > 0) {
    env.CLAUDE_CODE_ADDITIONAL_DIRECTORIES_CLAUDE_MD = "1";
    env.VIBE_ADDITIONAL_REPOS = JSON.stringify(additionalRepos);
  }
  if (targetBranch) {
    env.VIBE_BASE_BRANCH = targetBranch;
  }

  const proc = Bun.spawn(args, {
    cwd: worktreePath,
    stdin: "inherit",
    stdout: "inherit",
    stderr: "inherit",
    env,
  });
  await proc.exited;
}

async function addToClaudeConfig(home: string, worktreePath: string): Promise<void> {
  const configPath = join(home, ".claude.json");
  const configFile = Bun.file(configPath);

  try {
    if (!(await configFile.exists())) return;

    const config = await configFile.json() as Record<string, unknown>;
    const projects = (config.projects as Record<string, unknown>) ?? {};

    if (!(worktreePath in projects)) {
      projects[worktreePath] = { hasTrustDialogAccepted: true };
      config.projects = projects;
      await Bun.write(configPath, JSON.stringify(config, null, 2) + "\n");
    }
  } catch {
    // Ignore errors
  }
}
